### PR TITLE
feat(ci): Introduce 'regression detection' flow

### PR DESCRIPTION
**Pull request by blt**
This pull request introduces the 'regression detection' flow, a more advanced version of the existing 'soak' flow. The basic idea is the same: use a carefully controlled experimental setup to make a statistical argument about the performance impact of small changes. The regression detector has a proper team behind it now -- Single Machine Performance -- and has a more effective statistical model behind it.

This PR lacks analysis. This is expected. SMP does not yet have analysis working in our private compute environment -- please chat with @goxberry for details -- but we will soon. That said, it's very helpful for us to have the regression detector running on live PRs to understand if we've got your cluster right-sized etc.

We will add analysis via PR in the future. For now, the regression detector runs in parallel to the existing soaks but does not compete with them for compute resources in any way.

Signed-off-by: Brian L. Troutwine <brian.troutwine@datadoghq.com>

### CODE REVIEW

File: .github/workflows/regression.yml [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+  confirm-valid-credentials:
+    name: Confirm AWS credentials are minimally valid
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_ACCESS_KEY }}
```
Comment (jszwedko) [inline .github/workflows/regression.yml:43]:
```suggestion
          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_SECRET_ACCESS_KEY }}
```

Nit.

File: .github/workflows/regression.yml [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2.2.1
+
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_ACCESS_KEY }}
```
Comment (jszwedko) [inline .github/workflows/regression.yml:158]:
```suggestion
          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_SECRET_ACCESS_KEY }}
```

File: .github/workflows/regression.yml [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
```
Comment (jszwedko) [inline .github/workflows/regression.yml:206]:
```suggestion
          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_SECRET_ACCESS_KEY }}
```

File: .github/workflows/regression.yml [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+      - build-baseline
+      - build-comparison
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_ACCESS_KEY }}
```
Comment (jszwedko) [inline .github/workflows/regression.yml:248]:
```suggestion
          aws-secret-access-key: ${{ secrets.SINGLE_MACHINE_PERFORMANCE_BOT_SECRET_ACCESS_KEY }}
```

**Review by jszwedko — COMMENTED** [review]
Nice and tidy!

My only concern is that this workflow won't run on contributor PRs due to lack of access to secrets. I think it is valuable to have the regression tests run on contributor PRs so I think our options are:

* Use `pull_request_target` as the trigger which will run the workflow as it exists in the base of the pull request. We'd want to avoid checking out the PR code in the workflow though, to avoid compromising the secrets, so this workflow would need to be reworked a bit to split up the bits that depend on the code in the PR to a separate workflow that is triggered by `pull_request`. See https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
* Move this workflow to the internal Gitlab. In general, I like having the workflows in Github Actions because it enables others to contribute to them, but outside contributors won't really be able to meaningfully contribute to this workflow anyway.

**Comment by blt** [comment]
> Nice and tidy!
> 
> My only concern is that this workflow won't run on contributor PRs due to lack of access to secrets. I think it is valuable to have the regression tests run on contributor PRs so I think our options are:
> 
>     * Use `pull_request_target` as the trigger which will run the workflow as it exists in the base of the pull request. We'd want to avoid checking out the PR code in the workflow though, to avoid compromising the secrets, so this workflow would need to be reworked a bit to split up the bits that depend on the code in the PR to a separate workflow that is triggered by `pull_request`. See https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
> 
>     * Move this workflow to the internal Gitlab. In general, I like having the workflows in Github Actions because it enables others to contribute to them, but outside contributors won't really be able to meaningfully contribute to this workflow anyway.

Taking a run at `pull_request_target` is my preference here. I think internal Gitlab is fine but on account of this being an open project I'd prefer not to have totally secretive CI bits if it can be helped.

**Comment by spencergilbert** [comment]
> Nice and tidy!
> 
> My only concern is that this workflow won't run on contributor PRs due to lack of access to secrets. I think it is valuable to have the regression tests run on contributor PRs so I think our options are:
> 
> * Use `pull_request_target` as the trigger which will run the workflow as it exists in the base of the pull request. We'd want to avoid checking out the PR code in the workflow though, to avoid compromising the secrets, so this workflow would need to be reworked a bit to split up the bits that depend on the code in the PR to a separate workflow that is triggered by `pull_request`. See https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
> * Move this workflow to the internal Gitlab. In general, I like having the workflows in Github Actions because it enables others to contribute to them, but outside contributors won't really be able to meaningfully contribute to this workflow anyway.

I don't think this strictly blocks this PR, as it's still being run alongside the existing soaks. I'm not sure how this could be moved to GitLab _and_ kept in this repo, but I did raise this to @blt a week or two ago and I think he'd opened a Jira issue?

**Review by spencergilbert — APPROVED** [review]
Jesse's naming nits aside (which I agree with), I'm happy to have this first iteration land to help with sizing and continued development of the service.

I don't think there's a good, quick answer for secret management but I don't think it's a hard requirement for this initial step - it is a requirement for the service and eventual replacement of our in-repo soaks.

**Comment by blt** [comment]
> Jesse's naming nits aside (which I agree with), I'm happy to have this first iteration land to help with sizing and continued development of the service.
> 
> I don't think there's a good, quick answer for secret management but I don't think it's a hard requirement for this initial step - it is a requirement for the service and eventual replacement of our in-repo soaks.

After fiddling with the `pull_request`, `pull_request_target` approach some and talking to Jessie and building off Spencer's comments I think here's my appetite:

* move forward as-is (updating the secret names per nit)
* allow @goxberry to add analysis in the future in the current flow that contributors cannot run 
* once the flow works end-to-end to the point of being able to comment, consider a `pull_request`, `pull_request_target` split approach 

From the Single Machine Performance side of things our goal in the future is to just occasionally need to update the `smp` binary -- @GeorgeHahn is thinking most about our release process. Once the flow is generally understood that leaves us more or less okay with chunks migrating to `pull_request_target` or even `workflow_run` flows, since the need to merge these into master before changes kick in is not a real problem anymore.

**Review by jszwedko — APPROVED** [review]

**Pull request merged**